rama-net/src/address/socket_address.rs

```diff
@@ -0,0 +1,295 @@
+    type Error = OpaqueError;
+
+    fn try_from(s: &str) -> Result<Self, Self::Error> {
+        let (ip_addr, port) = crate::address::authority::split_port_from_str(s)?;
+        let ip_addr = try_to_parse_str_to_ip(ip_addr).context("parse ip address from socket address")?;
+        match ip_addr {
+            IpAddr::V6(_) if !s.starts_with('[') => Err(OpaqueError::from_display(
+                "missing brackets for IPv6 address with port",
+            )),
+            _ => Ok(SocketAddress { ip_addr, port }),
+        }
+    }
+}
+
+#[cfg(feature = "http")]
+impl TryFrom<HeaderValue> for SocketAddress {
+    type Error = OpaqueError;
+
+    fn try_from(header: HeaderValue) -> Result<Self, Self::Error> {
+        Self::try_from(&header)
+    }
+}
+
+#[cfg(feature = "http")]
+impl TryFrom<&HeaderValue> for SocketAddress {
+    type Error = OpaqueError;
+
+    fn try_from(header: &HeaderValue) -> Result<Self, Self::Error> {
+        header.to_str().context("convert header to str")?.try_into()
+    }
+}
+
+impl TryFrom<Vec<u8>> for SocketAddress {
+    type Error = OpaqueError;
+
+    fn try_from(bytes: Vec<u8>) -> Result<Self, Self::Error> {
+        Self::try_from(bytes.as_slice())
+    }
+}
+
+impl TryFrom<&[u8]> for SocketAddress {
+    type Error = OpaqueError;
+
+    fn try_from(bytes: &[u8]) -> Result<Self, Self::Error> {
+        let s = std::str::from_utf8(bytes).context("parse sock address from bytes")?;
+        s.try_into()
+    }
+}
+
+impl serde::Serialize for SocketAddress {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let address = self.to_string();
+        address.serialize(serializer)
+    }
+}
+
+impl<'de> serde::Deserialize<'de> for SocketAddress {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let s = String::deserialize(deserializer)?;
+        s.try_into().map_err(serde::de::Error::custom)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    fn assert_eq(s: &str, sock_address: SocketAddress, ip_addr: &IpAddr, port: u16) {
+        assert_eq!(
+            sock_address.ip_addr(),
+            ip_addr,
+            "parsing: {}",
+            s
+        );
+        assert_eq!(sock_address.port(), port, "parsing: {}", s);
+    }
+
+    #[test]
+    fn test_parse_valid() {
+        for (s, expected_socket_address) in [
+            ("[::1]:80", SocketAddress::new("::1".parse().unwrap(), 80)),
+            ("127.0.0.1:80", SocketAddress::new("127.0.0.1".parse().unwrap(), 80)),
+            (
+                "[2001:db8:3333:4444:5555:6666:7777:8888]:80",
+                SocketAddress::new("2001:db8:3333:4444:5555:6666:7777:8888".parse().unwrap(), 80),
+            ),
+        ] {
+            let msg = format!("parsing '{}'", s);
+
+            assert_eq(s, s.parse().expect(&msg), expected_socket_address.ip_addr(), expected_socket_address.port());
```

Comment:
```suggestion
            assert_eq(expected_socket_address, s.parse().expect(&msg), expected_socket_address);
```

Not sure if it will compile, but this is more or less the change you need to make for also all the assert_eq's below this. Goal is to compare to an actual address, otherwise you're just validating string with string, which again to be fair is also due to my fault of having that buggy test in the first place in `authority.rs`